osutil/writer.go

```diff
@@ -0,0 +1,64 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"fmt"
+	"io"
+)
+
+type BufferLimitError struct{}
+
+func (e BufferLimitError) Error() string {
+	return fmt.Sprintf("buffer capacity exceeded")
+}
+
+type LimitedWriter struct {
+	writer             io.Writer
+	limit              int
+	limitReached       bool
+	written            int
+	BufferLimitReached chan struct{}
+}
+
+func NewLimitedWriter(writer io.Writer, limit int) *LimitedWriter {
+	return &LimitedWriter{
+		writer:             writer,
+		limit:              limit,
+		BufferLimitReached: make(chan struct{}),
+	}
+}
+
+func (wr *LimitedWriter) Write(data []byte) (int, error) {
+	n := len(data) + wr.written
+	if n > wr.limit {
+		// in case Write is called after error - closing channel again
+		// would panic
+		if !wr.limitReached {
+			close(wr.BufferLimitReached)
+		}
+		wr.limitReached = true
+		return 0, &BufferLimitError{}
```

Comment:
I apologize for not suggesting this earlier, but I suggest we go a slightly different route here, which will be much more stable and more appreciated by people as well. Reading this I got concerned about people calling external processes that they don't realize can general a lot of data on exceptional cases (think apt), and then getting devices dying in production because of that.

So instead of killing the process, let's implement a _LimitedBuffer_ where you can write as much data on it as you want, but only so many lines and up to max bytes will come out of its Bytes() method. To implement it efficiently, we can use a bytes.Buffer internally, and write on it up up to maxBytes*2. When we hit that limit during Write, we call that same truncate function, and copy the resulting range over the original buffer (in effect, copy a range close to the end of the buffer at the start of itself).

That logic goes on Write. On Bytes, we simply call truncate once and return the result.

How does that sound?

Reply:
Ah, LimitedBuffer should live in strutil as well, not osutil.

Reply:
That's a valid concern, no objections. I'll redo this, thanks for the review!